File: packages/lightning/src/commands/bridge/join.ts
import type { command_opts } from '../../structures/commands.ts';
import { log_error } from '../../structures/errors.ts';
import { bridge_add_common } from './_internal.ts';

export async function join(
	opts: command_opts,
): Promise<string> {
	const result = await bridge_add_common(opts);

	if (typeof result === 'string') return result;

	const target_bridge = await opts.lightning.data.get_bridge_by_id(
		opts.args.id,
	);

	if (!target_bridge) {
		return `Bridge with id \`${opts.args.id}\` not found. Make sure you have the correct id.`;
	}

	target_bridge.channels.push(result);

	try {
		await opts.lightning.data.edit_bridge(target_bridge);

		return `Bridge joined successfully!`;
	} catch (e) {
		log_error(e, {
			message: 'Failed to update bridge in database',
			extra: { target_bridge },
		});
	}
}

File: internal/commands/join.go
package commands

import (
	"github.com/williamhorning/lightning/internal/data"
	"github.com/williamhorning/lightning/pkg/lightning"
)

func bridgeJoin(database data.Database, name string) *lightning.Command {
	cmd := &lightning.Command{
		Name:        name,
		Description: "join an existing bridge with the given ID",
		Arguments:   []*lightning.CommandArgument{{Name: "id", Description: "the bridge id to use", Required: true}},
		Executor: func(opts *lightning.CommandOptions) {
			if opts.Arguments["id"] == "" {
				sendErr(missingArgumentError{argument: "id"}, "missing argument", opts)

				return
			}

			bridge, err := database.GetBridge(opts.Arguments["id"])
			if err != nil {
				sendErr(err, "failed to get bridge by id", opts)

				return
			} else if bridge.ID == "" {
				sendErr(bridgeNotFoundError{}, "no bridge with that ID exists", opts)

				return
			}

			channel, err := prepareChannelForBridge(database, opts)
			if err != nil {
				sendErr(err, "failed to setup channel for bridge", opts)

				return
			}

			channel.Disabled = lightning.ChannelDisabled{Read: name == "subscribe", Write: false}
			bridge.Channels = append(bridge.Channels, *channel)

			if err = database.CreateBridge(bridge); err != nil {
				sendErr(err, "failed to update bridge in the database", opts)

				return
			}

			if err = opts.Reply(
				getMessage("joined bridge!", "you successfully joined the bridge `"+bridge.ID+"`!"), true,
			); err != nil {
				sendErr(err, "failed to respond to command", opts)
			}
		},
	}

	if name == "subscribe" {
		cmd.Description = "subscribe to an existing bridge with the given ID (read-only)"
	}

	return cmd
}
